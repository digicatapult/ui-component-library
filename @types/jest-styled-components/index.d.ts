declare module 'jest-styled-components/serializer'
