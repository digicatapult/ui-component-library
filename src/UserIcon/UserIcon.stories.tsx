import React from 'react'
import { Story, StoryObj } from '@storybook/react'

import UserIcon from './index.js'
import { styled } from 'styled-components'

export default {
  title: 'Components/UserIcon',
  component: UserIcon,
  argTypes: {
    name: {
      control: 'text',
    },
    size: {
      control: 'text',
    },
    color: {
      control: 'color',
    },
    bgColor: {
      control: 'color',
    },
    outlineColor: {
      control: 'color',
    },
  },
}

export const Default: StoryObj = {
  args: {},
}
export const HII: StoryObj = {
  render: (args) => (
    <Wrapper>
      <UserIcon {...args} />
    </Wrapper>
  ),
  args: {
    name: 'bruce wayne',
    size: '70px',
    color: '#1a1a1a',
    bgColor: '#9edcfa',
    outlineColor: 'white',
  },
}

const Wrapper = styled.div`
  background: #1a1a1a;
  padding: 16px;
`
