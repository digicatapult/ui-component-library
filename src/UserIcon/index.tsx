import React from 'react'
import styled from 'styled-components'

export type Props = {
  fullName?: string
  size?: string
  color?: string
  bgColor?: string
  outlineColor?: string
}

const defaults: Required<Props> = {
  fullName: 'A A',
  size: '4em',
  color: '#efefef',
  bgColor: '#1E5CbA',
  outlineColor: '#101010',
}

const getInitialsFromName = (fullName: string) => {
  const allInitials = !!fullName
    ? fullName
        .split(' ')
        .filter((s) => !!s)
        .map((s) => s[0])
    : []

  // 3 case: no initials, 1 initial, more than 1 initial
  switch (allInitials.length) {
    case 0:
      return 'AA'
    case 1:
      return allInitials[0]
    default:
      return allInitials[0] + allInitials[allInitials.length - 1]
  }
}

const UserIcon: React.FC<Props> = (props) => {
  const withDefaults = { ...defaults, ...props }
  const initials = getInitialsFromName(withDefaults.fullName)
  return (
    <Wrapper {...withDefaults}>
      <p>{initials}</p>
    </Wrapper>
  )
}

const Wrapper = styled.div<Required<Props>>`
  box-sizing: border-box;
  display: grid;
  align-items: center;
  justify-content: center;
  width: ${({ size }) => size};
  font-size: calc(${({ size }) => size} / 4);
  text-transform: uppercase;
  aspect-ratio: 1 / 1;
  border-radius: 50%;
  border: calc(${({ size }) => size} * 0.065) solid
    ${({ outlineColor }) => outlineColor};
  color: ${({ color }) => color};
  background: ${({ bgColor }) => bgColor};
`

export default UserIcon
