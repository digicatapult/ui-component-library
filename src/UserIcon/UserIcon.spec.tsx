import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import UserIcon from './index.js'

describe('Link', () => {
  test('default', () => {
    const tree = renderer.create(<UserIcon />).toJSON()
    expect(tree).toMatchSnapshot()
  })
  test('overrides', () => {
    const tree = renderer
      .create(
        <UserIcon
          fullName={'bruce wayne'}
          size={'70px'}
          color={'#1a1a1a'}
          bgColor={'#9edcfa'}
          outlineColor={'white'}
        />,
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
