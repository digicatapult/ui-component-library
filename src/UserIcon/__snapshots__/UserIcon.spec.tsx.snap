// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link default 1`] = `
.c0 {
  box-sizing: border-box;
  display: grid;
  align-items: center;
  justify-content: center;
  width: 4em;
  font-size: calc(4em / 4);
  text-transform: uppercase;
  aspect-ratio: 1/1;
  border-radius: 50%;
  border: calc(4em * 0.065) solid #101010;
  color: #efefef;
  background: #1E5CbA;
}

<div
  bgColor="#1E5CbA"
  className="c0"
  color="#efefef"
  fullName="A A"
  outlineColor="#101010"
  size="4em"
>
  <p>
    AA
  </p>
</div>
`;

exports[`Link overrides 1`] = `
.c0 {
  box-sizing: border-box;
  display: grid;
  align-items: center;
  justify-content: center;
  width: 70px;
  font-size: calc(70px / 4);
  text-transform: uppercase;
  aspect-ratio: 1/1;
  border-radius: 50%;
  border: calc(70px * 0.065) solid white;
  color: #1a1a1a;
  background: #9edcfa;
}

<div
  bgColor="#9edcfa"
  className="c0"
  color="#1a1a1a"
  fullName="bruce wayne"
  outlineColor="white"
  size="70px"
>
  <p>
    bw
  </p>
</div>
`;
